working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+
+*The show/hide part provides access to a subset of the members of the
+interface of the on-type. For instance, there may be some read-only methods
+that we can safely call on the on-type, because they won't violate any
+invariants associated with the extension type. We could write forwarding
+members in the extension body, but using show/hide can have the same
+effect, and it is much more concise and convenient.*
+
+We use the phrase _extension show/hide part_, or just _show/hide part_ when
+no doubt can arise, to denote a phrase derived from
+`<extensionShowHidePart>`. Similarly, an `<extensionShowClause>` is known
+as an _extension show clause_, and an `<extensionHideClause>` is known as
+an _extension hide clause_, similarly abbreviated to _show clause_ and
+_hide clause_.
+
+The show/hide part specifies which instance members of the on-type are
+available for invocation on a receiver whose type is the given extension
+type.
+
+If the show/hide part is empty, no instance members except the ones
+declared for `Object?` can be invoked on a receiver whose static type is
+the given extension type.
+
+If the show/hide part is a show clause listing some identifiers and types,
+invocation of an instance member is allowed if its basename is one of the
+given identifiers, or it is the name of a member of the interface of one of
+the types. Instance members declared for `Object?` can also be invoked.
+
+If the show/hide part is a hide clause listing some identifiers and types,
+invocation of an instance member is allowed if it is in the interface of
+the on-type and _not_ among the given identifiers, nor in the interface of
+the specified types.
+
+If the show/hide part is a show clause followed by a hide clause, then the
+available instance members is computed by first computing the set of
+included instance members specified by the show clause as described above,
+and then removing instance members from that set according to the hide
+clause, as described above.
+
+In a show or hide clause, it is possible that an
+`<extensionShowHideElement>` is an identifier that is the basename of a
+member of the interface of the on-type, and it is also the name of a type
+in scope. In this case, the name shall refer to the member.
+
+*A conflict is unlikely because type names in general are capitalized, and
+member names start with a lower-case letter. Some type names start with a
+lower-case letter, too (e.g., `int` and `dynamic`), but those names do not
+occur frequently as member names. Should a conflict arise anyway, a
+work-around is to import the shadowed type `T` with a prefix `p` and put
+`p.T` in the show or hide clause.*
+
+A compile-time error occurs if a hide or show clause contains an identifier
+which is not the basename of an instance member of the on-type. A
+compile-time error occurs if a hide or show clause contains a type which is
+not among the types that are implemented by the on-type of the extension.
```

Comment:
This is probably ok for now, but we will need to define what "implemented by" means.  What are the types "implemented by" an extension type?  By `FutureOr`?  By a nullable type?  By a generic type with a bound?  By a function type?

Reply:
The main point here is that the "implements" requirement ensures that each item in the hide/show part is relevant (e.g., such that we don't say `hide Widget` when the on-type is `int`). We could easily relax this restriction if we see a need.

About the phrase 'a type `S` implements another type `T`': It is [defined to hold](https://github.com/dart-lang/language/blob/53f3e4606f7efe25679c37698e167734df7f934c/specification/dartLangSpec.tex#L4975) when there is a superinterface relation.

So `List<T>` implements `Iterable<T>` for any `T`, but `List<T>` does not implement `Iterable<Object>` for any `T` other than `Object`; or `T` is a raw type such that there exist actual type arguments `T1,.. Tk` such that `S` implements `T<T1, .. Tk>`. (In the latter case `T1, .. Tk` are uniquely determined, so there's no loss of information.)

The types implemented by an extension type would need to be defined. For the purpose of this paragraph it would work fine to say that an extension type `E` with `implements T1, .. Tk` implements `T1, .. Tk`.

I don't see anything in the section about `FutureOr` that justifies a claim that any particular type is a superinterface of `FutureOr`. So `FutureOr` doesn't implement any other type.

Similarly for a type of the form `T?`: It doesn't implement anything.

I believe that the class `Null` implements `Object` (the spec mentions that it 'extends' `Object`).

In summary, I don't really think it's a significant inconvenience (and it does seem to be a reasonable sanity check) that we require the show/hide part to use types that are superinterfaces (rather than arbitrary types).